### ISSUE TITLE
Support resource owner password flow.

### ISSUE THREAD

**Issue by aseif83** [FormidableLabs/react-native-app-auth#488]
I want to use the this package to implement resource owner password credential grant (aka password) in my new App.

**Comment by kadikraman** [comment]
Sorry, we won't be supporting password grant. The latest OAuth 2.0 Security Best Current Practice disallows the password grant entirely - [source](https://oauth.net/2/grant-types/password/).

**Comment by janpieterz** [comment]
@kadikraman We're running into an "issue" where automated tests running in the AppCenter device testing suite won't be able to interact with the Google Chrome / Safari views. This means that running automated tests locally is fine (you can log-in once) but running them on their suite of devices breaks as soon as you try to log in.

For this reason we're looking to use a hidden password grant method, as we'd love to stay as close to the real implementation as possible (for example ensuring things like keychain interaction, the logic of handling RN app auth results etc stay the same). 

Any thoughts on this? 
Any RN library you're aware of that implements password grants easily, or an easy way to hack it in there with RN App Auth to prevent a dependency for testing only?

**Comment by kadikraman** [comment]
With the password grant, your application would collect the username/password and use it to do a POST request to the token endpoint (effectively skipping the whole part with opening a webview and asking the user to log in via 3rd party).

Since a huge part of what this library does is handling the 3rd party login, I don't think there's an easy way to mock a password grant, but if I'm reading [these](https://developer.okta.com/blog/2018/06/29/what-is-the-oauth2-password-grant) [articles](https://www.oauth.com/oauth2-servers/access-tokens/password-grant/) correctly, given that your IDP supports it, you can authenticate by using the password grant and hardcoding the username/password in your tests, and doing a POST request to the token endpoint.

```
POST /oauth/token HTTP/1.1
Host: authorization-server.com
 
grant_type=password
&username=user@example.com
&password=1234luggage
&client_id=xxxxxxxxxx
&client_secret=xxxxxxxxxx
```

The only difference between that response and what we return is the formatting, i.e

```objective-c
/*
 * Take raw OIDTokenResponse and turn it to a token response format to pass to JavaScript caller
 */
- (NSDictionary*)formatResponse: (OIDTokenResponse*) response {
    NSDateFormatter *dateFormat = [[NSDateFormatter alloc] init];
    dateFormat.timeZone = [NSTimeZone timeZoneWithAbbreviation: @"UTC"];
    [dateFormat setLocale:[NSLocale localeWithLocaleIdentifier:@"en_US_POSIX"]];
    [dateFormat setDateFormat:@"yyyy-MM-dd'T'HH:mm:ss'Z'"];

    return @{@"accessToken": response.accessToken ? response.accessToken : @"",
             @"accessTokenExpirationDate": response.accessTokenExpirationDate ? [dateFormat stringFromDate:response.accessTokenExpirationDate] : @"",
             @"additionalParameters": response.additionalParameters,
             @"idToken": response.idToken ? response.idToken : @"",
             @"refreshToken": response.refreshToken ? response.refreshToken : @"",
             @"tokenType": response.tokenType ? response.tokenType : @"",
             };
}

```

Basically we can't mock it with app auth, but you shouldn't need another library, I think it's just one `post` request so it can be done with `fetch`.

**Comment by janpieterz** [comment]
Thanks @kadikraman , it makes sense. I think we can capture this just before entering into RN-app-auth and then continuing further as if everything is the same. Thanks a lot for the deep-dive here.
